jetbrains-core/src/software/aws/toolkits/jetbrains/services/telemetry/TelemetryService.kt

```diff
@@ -0,0 +1,160 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.services.telemetry
+
+import com.intellij.openapi.Disposable
+import com.intellij.openapi.application.ApplicationInfo
+import com.intellij.openapi.application.ApplicationNamesInfo
+import com.intellij.openapi.components.ServiceManager
+import com.intellij.openapi.project.DefaultProjectFactory
+import com.intellij.openapi.util.SystemInfo
+import software.amazon.awssdk.services.toolkittelemetry.ToolkitTelemetryClient
+import software.amazon.awssdk.services.toolkittelemetry.model.AWSProduct
+import software.amazon.awssdk.services.toolkittelemetry.model.Unit
+import software.aws.toolkits.core.ToolkitClientManager
+import software.aws.toolkits.core.region.ToolkitRegionProvider
+import software.aws.toolkits.core.telemetry.MetricEvent
+import software.aws.toolkits.core.telemetry.TelemetryBatcher
+import software.aws.toolkits.core.telemetry.AWSCognitoCredentialsProvider
+import software.aws.toolkits.core.telemetry.DefaultTelemetryBatcher
+import software.aws.toolkits.core.telemetry.DefaultTelemetryPublisher
+import software.aws.toolkits.core.telemetry.DefaultMetricEvent
+import software.aws.toolkits.core.utils.getLogger
+import software.aws.toolkits.jetbrains.AwsToolkit
+import software.aws.toolkits.jetbrains.core.AwsSdkClient
+import software.aws.toolkits.jetbrains.settings.AwsSettings
+import java.net.URI
+import java.time.Duration
+import java.time.Instant
+import java.util.concurrent.Executors
+import java.util.concurrent.ScheduledExecutorService
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicBoolean
+
+interface TelemetryService : Disposable {
+    fun record(buildEvent: MetricEvent.Builder.() -> kotlin.Unit): MetricEvent
+}
+
+class DefaultTelemetryService(
+    sdkClient: AwsSdkClient,
+    regionProvider: ToolkitRegionProvider,
+    messageBusService: MessageBusService,
+    publishInterval: Long = DEFAULT_PUBLISH_INTERVAL,
+    publishIntervalUnit: TimeUnit = DEFAULT_PUBLISH_INTERVAL_UNIT,
+    private val executor: ScheduledExecutorService = createDefaultExecutor(),
+    batcher: TelemetryBatcher? = null
+) : TelemetryService {
+    private val isDisposing: AtomicBoolean = AtomicBoolean(false)
+    private val startTime: Instant
+    private val _batcher: TelemetryBatcher
+
+    init {
+        _batcher = batcher ?: {
+            val settings = AwsSettings.getInstance()
+            val clientManager = ServiceManager.getService(
+                    DefaultProjectFactory.getInstance().defaultProject,
+                    ToolkitClientManager::class.java
+            )
+            val client: ToolkitTelemetryClient = ToolkitTelemetryClient
+                    .builder()
+                    .endpointOverride(URI.create("https://client-telemetry.us-east-1.amazonaws.com"))
+                    // TODO: Determine why this client is not picked up by default.
```

Comment:
This is probably fixed by this: https://github.com/aws/aws-sdk-java-v2/commit/87fe163d3b563c3160161cd451f25b52d67c5b4e

Reply:
We can test/fix this at a later date

Reply:
Unfortunately, this doesn't appear to be fixed. Leaving as-is for now.